core/chains/evm/chain_set.go

```diff
@@ -91,14 +105,112 @@ func (cll *chainSet) Default() (Chain, error) {
 	return cll.Get(cll.defaultID)
 }
 
+// Requires a lock on chainsMu
+func (cll *chainSet) initializeChain(dbchain *types.Chain) error {
+	// preload nodes
+	// TODO: replace with math.MaxInt once we make go 1.17 mandatory
+	nodes, _, err := cll.orm.NodesForChain(dbchain.ID, 0, math.MaxInt16)
+	if err != nil {
+		return err
+	}
+	dbchain.Nodes = nodes
+
+	cid := dbchain.ID.String()
+	chain, err := newChain(*dbchain, cll.opts)
+	if errors.Cause(err) == ErrNoPrimaryNode || len(dbchain.Nodes) == 0 {
+		cll.logger.Warnf("EVM: No primary node found for chain %s; this chain will be ignored", cid)
+		return nil
+	} else if err != nil {
+		return err
+	}
+	if err = chain.Start(); err != nil {
+		return err
+	}
+	cll.chains[cid] = chain
+	return nil
+}
+
+func (cll *chainSet) Add(id *big.Int, config types.ChainCfg) (types.Chain, error) {
+	cll.chainsMu.Lock()
+	defer cll.chainsMu.Unlock()
+
+	cid := id.String()
+	if _, exists := cll.chains[cid]; exists {
+		return types.Chain{}, errors.Errorf("chain already exists with id %d", id)
```

Comment:
Golang is retarded and I swear I have seen that if you `%d` a bigint you can get a garbage number that isn't the value at all (although I cannot reproduce on the playground).

I usually do `%s`, x.String() which seems the safest